From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alex Cameron <asc@tetsuo.sh>
Date: Wed, 3 Aug 2022 20:01:18 +1000
Subject: [PATCH] 4497: Use `stroull` instead of `stroul` to parse address
 offsets

Check endptr and `errno` in error checking

Set `errno` to 0 prior to the `stroull` call
---
 Ghidra/Features/Decompiler/src/decompile/cpp/space.cc | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/space.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/space.cc
index 40a3e3bfe..a921aee16 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/space.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/space.cc
@@ -16,6 +16,8 @@
 #include "space.hh"
 #include "translate.hh"
 
+#include <climits>
+
 AttributeId ATTRIB_BASE = AttributeId("base",89);
 AttributeId ATTRIB_DEADCODEDELAY = AttributeId("deadcodedelay",90);
 AttributeId ATTRIB_DELAY = AttributeId("delay", 91);
@@ -268,7 +270,11 @@ uintb AddrSpace::read(const string &s,int4 &size) const
     }
   }
   catch(LowlevelError &err) {	// Name doesn't exist
-    offset = strtoul(s.c_str(),&tmpdata,0);
+    errno = 0;
+    offset = strtoull(s.c_str(),&tmpdata,0);
+    if (tmpdata == s.c_str() || *tmpdata != '\0' || (offset == ULLONG_MAX && errno == ERANGE)) {
+      throw LowlevelError("Offset outside of valid range");
+    }
     offset = addressToByte(offset,wordsize);
     enddata = (const char *) tmpdata;
     if (enddata - s.c_str() == s.size()) { // If no size or offset override
-- 
2.38.1

